content/en/blog/_posts/2023-12-05-image-filesystem.md

```diff
@@ -0,0 +1,171 @@
+graphroot = "/var/lib/containers/storage"
+```
+
+- graphroot
+  - Persistent data stored from the container runtime
+  - If SELinux is enabled, this must match the `/var/lib/containers/storage`
+- runroot
+  - Temporary read/write access for container.
+  - Recommended to have this on a temporary filesystem.
+
+A quick way to relabel your graphroot directory to match `/var/lib/containers/storage`.
+
+```bash
+semanage fcontext -a -e /var/lib/containers/storage /YOUR_STORAGE_PATH
+restorecon -R -v /YOUR_STORAGE_PATH
+```
+
+### Containerd
+
+containerd uses a toml configuration file to control where persistent and ephemeral data is stored.
+The default path for the config file is located at `/etc/containerd/config.toml`.
+
+The relevant fields for containerd storage are `root` and `state`.
+
+- root
+  - The root directory for containerd metadata
+  - Default is `/var/lib/containerd`
+  - Root also requires selinux labels if your OS requires it
+- state
+  - Temporary data for containerd
+  - Default is `/run/containerd`
+
+## Kubernetes Node Pressure Eviction
+
+Kubernetes will automatically detect if the container filesystem is split from the node filesystem. When one separates the filesystem, Kubernetes is responsible for monitoring both the node filesystem and the container runtime filesystem.
+Kubernetes documentation refers to the node filesystem and the container runtime filesystem as nodefs and imagefs.
+If either nodefs or the imagefs are running out of disk space, then the overall node is considered to have disk pressure.
+Kubernetes will first reclaim space by deleting unusued containers and images, and then it will resort to evicting pods.
+If one has a imagefs, then Kubernetes will garbage collect unused images on imagefs and will remove dead containers and pods from the nodefs.
+If there is only a nodefs, then Kubernetes garbage collection includes dead containers, dead pods and unused images.
+
+Kubernetes allows more configurations for determining if your disk is full.  
+The eviction manager in Kubelet defines API fields to signal pressure.
+For filesystems, the relevant API are `nodefs.available`, `nodefs.inodesfree`, `imagefs.available`, and `imagefs.inodesfree`.
+If there is no container runtime filesystem then imagefs is ignored.
+
+Users can use the existing defaults:
+
+- memory.available<100Mi
+- nodefs.available<10%
+- imagefs.available<15%
+- nodefs.inodesFree<5% (Linux nodes)
+
+Kubernetes allows one to set user defined values in `EvictionHard` and `EvictionSoft` in the Kubelet Configuration.  
+
+EvictionHard defines limits and once these limits are exceeded, pods will be evicted without any grace period.
+EvictionSoft defines limits and once these limits are exceeded, pods will be evicted with a grace period that can be set per signal.
+
+If you specify a `EvictionHard` it will replace the defaults.  This means it is important to set all signals in your configuration.
```

Comment:
```suggestion
If you specify an `EvictionHard`, it will replace the defaults.  This means it is important to set all signals in your configuration.
```